src/components/Composer/index.js

```diff
@@ -122,380 +123,352 @@ const IMAGE_EXTENSIONS = {
+        textInput.current.style.height = 0;
+        const computedStyle = window.getComputedStyle(textInput.current);
+        const lineHeight = parseInt(computedStyle.lineHeight, 10) || 20;
+        const paddingTopAndBottom = parseInt(computedStyle.paddingBottom, 10) + parseInt(computedStyle.paddingTop, 10);
+        setTextInputWidth(computedStyle.width);
+
+        const computedNumberOfLines = ComposerUtils.getNumberOfLines(maxLines, lineHeight, paddingTopAndBottom, textInput.current.scrollHeight);
+        const generalNumberOfLines = computedNumberOfLines === 0 ? numberOfLinesProp : computedNumberOfLines;
+
+        onNumberOfLinesChange(generalNumberOfLines);
+        updateIsFullComposerAvailable({isFullComposerAvailable, setIsFullComposerAvailable}, generalNumberOfLines);
+        setNumberOfLines(generalNumberOfLines);
+        textInput.current.style.height = 'auto';
+    }, [value, maxLines, numberOfLinesProp, isComposerFullSize, onNumberOfLinesChange, isFullComposerAvailable, setIsFullComposerAvailable]);
+
+    const handleKeyPress = useCallback(
+        (e) => {
+            // Prevent onKeyPress from being triggered if the Enter key is pressed while text is being composed
+            if (!onKeyPress || isEnterWhileComposition(e)) {
+                return;
+            }
+            onKeyPress(e);
+        },
+        [onKeyPress],
+    );
 
     /**
-     * Check the current scrollHeight of the textarea (minus any padding) and
-     * divide by line height to get the total number of rows for the textarea.
+     * Set the TextInput Ref
+     *
+     * @param {Element} el
      */
-    updateNumberOfLines() {
-        // Hide the composer expand button so we can get an accurate reading of
-        // the height of the text input
-        this.props.setIsFullComposerAvailable(false);
-
-        // We have to reset the rows back to the minimum before updating so that the scrollHeight is not
-        // affected by the previous row setting. If we don't, rows will be added but not removed on backspace/delete.
-        this.setState({numberOfLines: 1}, () => {
-            const computedStyle = window.getComputedStyle(this.textInput);
-            const lineHeight = parseInt(computedStyle.lineHeight, 10) || 20;
-            const paddingTopAndBottom = parseInt(computedStyle.paddingBottom, 10) + parseInt(computedStyle.paddingTop, 10);
-            const computedNumberOfLines = ComposerUtils.getNumberOfLines(this.props.maxLines, lineHeight, paddingTopAndBottom, this.textInput.scrollHeight);
-            const numberOfLines = computedNumberOfLines === 0 ? this.props.numberOfLines : computedNumberOfLines;
-            updateIsFullComposerAvailable(this.props, numberOfLines);
-            this.setState({
-                numberOfLines,
-                width: computedStyle.width,
-            });
-            this.props.onNumberOfLinesChange(numberOfLines);
-        });
-    }
+    const setTextInputRef = useCallback(
```

Comment:
Why the logic of setting input ref has been changed? This could be done in `useEffect`, eventually `useLayoutEffect` as it was in componentDidMount originally.

Reply:
During the initial render, refs are null inside useEffect. To ensure we have a valid ref during the initial run of useEffect, I'm using a useCallback to set the ref from a child component

Reply:
During the initial render, refs are null inside useEffect. To ensure we have a valid ref during the initial run of useEffect, I'm using a useCallback to set the ref from a component

Reply:
I think callback is still not needed here, it could be `ref={(el) => (this.textInput = el)}`